target/linux/realtek/patches-5.10/320-irqchip-irq-realtek-use-hotplug.patch

```diff
@@ -0,0 +1,85 @@
+--- a/drivers/irqchip/irq-realtek-rtl.c
++++ b/drivers/irqchip/irq-realtek-rtl.c
+@@ -5,6 +5,7 @@
+  * Copyright (C) 2020 John Crispin <john@phrozen.org>
+  */
+ 
++#include <linux/cpuhotplug.h>
+ #include <linux/of_irq.h>
+ #include <linux/irqchip.h>
+ #include <linux/spinlock.h>
+@@ -118,7 +119,8 @@ static void realtek_ictl_mask_irq(struct
+ 	raw_spin_unlock_irqrestore(&irq_lock, flags);
+ }
+ 
+-static int __maybe_unused realtek_ictl_irq_affinity(struct irq_data *i,
++#ifdef CONFIG_SMP
++static int realtek_ictl_irq_affinity(struct irq_data *i,
+ 	const struct cpumask *dest, bool force)
+ {
+ 	struct realtek_ictl_output *output = i->domain->host_data;
+@@ -151,6 +153,22 @@ static int __maybe_unused realtek_ictl_i
+ 
+ 	return IRQ_SET_MASK_OK;
+ }
++#endif
++
++static int realtek_cpu_starting(unsigned int cpu)
++{	
++/*
++ * MIPS CPU startup function vsmp_init_secondary() will only enable some of
++ * the interrupts for the second CPU/VPE. Fix this during hotplug.
++ */  
++	if (cpu > 0)
++		change_c0_status(ST0_IM, 
++				STATUSF_IP0 | STATUSF_IP1 |
++				STATUSF_IP2 | STATUSF_IP3 |
++				STATUSF_IP4 | STATUSF_IP5 |
++				STATUSF_IP6 | STATUSF_IP7);
++	return 0;
++}
+ 
+ static struct irq_chip realtek_ictl_irq = {
+ 	.name = "realtek-rtl-intc",
+@@ -205,8 +223,8 @@ static int intc_select(struct irq_domain
+ 		routing_old = read_irr(REG(RTL_ICTL_IRR0, cpu), fwspec->param[0]);
+ 		routed_elsewhere = routing_old && fwspec->param[1] != routing_old - 1;
+ 		if (routed_elsewhere) {
+-			pr_warn("soc int %d already routed to output %d\n",
+-				fwspec->param[0], routing_old - 1);
++			pr_warn("%s soc int %d already routed to output %d\n",
++				__func__, fwspec->param[0], routing_old - 1);
+ 			break;
+ 		}
+ 	}
+@@ -316,7 +334,7 @@ static int __init realtek_rtl_of_init(st
+ 
+ 	num_parents = of_irq_count(node);
+ 	if (num_parents > RTL_ICTL_NUM_OUTPUTS) {
+-		pr_err("too many parent interrupts\n");
++		pr_err("%s too many parent interrupts\n", __func__);
+ 		return -EINVAL;
+ 	}
+ 
+@@ -352,6 +370,10 @@ static int __init realtek_rtl_of_init(st
+ 			return -ENODEV;
+ 	}
+ 
++	cpuhp_setup_state_nocalls(CPUHP_AP_IRQ_REALTEK_STARTING,
++				"irqchip/mips/realtek:starting",
++				realtek_cpu_starting, NULL);
++
+ 	return setup_parent_interrupts(node, &parent_irqs[0], num_parents);
+ }
+ 
+diff -Nurp a/include/linux/cpuhotplug.h b/include/linux/cpuhotplug.h
+--- a/include/linux/cpuhotplug.h	2022-09-02 08:41:35.013581830 +0200
++++ b/include/linux/cpuhotplug.h	2022-09-02 08:43:10.359405975 +0200
+@@ -106,6 +106,7 @@ enum cpuhp_state {
+ 	CPUHP_AP_IRQ_MIPS_GIC_STARTING,
+ 	CPUHP_AP_IRQ_RISCV_STARTING,
+ 	CPUHP_AP_IRQ_SIFIVE_PLIC_STARTING,
++	CPUHP_AP_IRQ_REALTEK_STARTING,
```

Comment:
Shouldn't this be sorted alphabetically, before `CPUHP_AP_IRQ_RISCV_STARTING`?

Reply:
I already checked but the whole timers list is not sorted at all.